File: problem2-bfs.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        indegrees = [0 for _ in range(numCourses)]
        adjList = {}
        for j in range(numCourses):
            adjList[j] = []
        for i in range(len(prerequisites)):
            indegrees[prerequisites[i][0]] +=1
            if prerequisites[i][1] in adjList:
                adjList[prerequisites[i][1]].append(prerequisites[i][0])
            else:
                adjList[prerequisites[i][1]] = ([prerequisites[i][0]]) 
                
        # print(adjList)
        # print(indegrees)
        q = []
        for i in range(len(indegrees)):
            if indegrees[i] == 0:
                q.append(i)
        if not q: return False
        count = 0
        while q:
            curr = q.pop(0)
            count +=1
            edges = adjList[curr]
            for edge in edges:
                indegrees[edge] -=1
                if indegrees[edge] == 0:
                    q.append(edge)
        
        if count!= numCourses: return False
        return True 

#Time complexity is O(v+e) where v is number of nodes and e is nomber of edges.
#Space complexity is O(v+e) as we are storing all the elements in auxiliary data strructure

File: Problem2_BFS.java
//Time Complexity: O(V+E), where v is the no of courses & e is the no of array elements in prerequisites.
//Space Complexity: O(V+E)
//Code run successfully on LeetCode.

public class Problem2_BFS {

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        
        if(numCourses ==0 || prerequisites == null)
            return false;
        
        
        int[] indegree = new int[numCourses];
        HashMap<Integer, List<Integer>> map = new HashMap<>();
        
        for(int[] pre : prerequisites){
            
            int in = pre[1];
            int out = pre[0];
            
            indegree[out]++;
            
            if(!map.containsKey(in))
                map.put(in, new ArrayList<>());
            
            map.get(in).add(out);
        }
        
        Queue<Integer> q = new LinkedList<>();
        
        for(int i =0; i < numCourses; i++){
            if(indegree[i] == 0)
                q.add(i);
        }
        
        while(!q.isEmpty()){
            
            int curr = q.poll();
        
            List<Integer> list = new ArrayList<>();
            list = map.get(curr);
            
            if(list != null){
                
                for(int edge : list){
                    
                    indegree[edge]--;
                    if(indegree[edge] == 0)
                        q.add(edge);
                }
            }
        }
        
        for(int i =0; i < numCourses; i++){
            
            if(indegree[i] != 0)
                return false;
        }
          
        return true;
    }
}
